modules/frontend/querysharding.go

```diff
@@ -167,3 +154,43 @@ func DoRequests(ctx context.Context, downstream Handler, reqs []*http.Request) (
 
 	return resps, firstErr
 }
+
+func mergeResponses(rrs []RequestResponse) (*http.Response, error) {
+	var errCode = http.StatusOK
+	var errBody io.ReadCloser
+	var combinedTrace []byte
+	for _, rr := range rrs {
+		if rr.Response.StatusCode == http.StatusOK {
+			body, err := ioutil.ReadAll(rr.Response.Body)
+			rr.Response.Body.Close()
+			if err != nil {
+				return nil, errors.Wrap(err, "error reading response body at query frontend")
+			}
+
+			if len(combinedTrace) == 0 {
+				combinedTrace = body
+			} else {
+				combinedTrace = util.CombineTraces(combinedTrace, body)
+			}
+		}
+		if rr.Response.StatusCode > errCode {
+			errCode = rr.Response.StatusCode
+			errBody = rr.Response.Body
+		}
+	}
+
+	// Always give precedence to 5xx errors and propagate them to the user so they can retry the query
+	if len(combinedTrace) == 0 || errCode > 499 {
```

Comment:
Should we translate all errorCodes > 400 (not 404) into 500?  Consider 400 (bad request), (401) unauthorized, etc...  Since we are building and executing the requests I think if we get 400 this should actually be a 500 externally.

This logic could use some ironing out for clarity:
```
errCode := 404
for _, rr := range rrs {
    if rr.statusCode = 200 {
      ...
    } else if rr.statusCode != 404 {
      errCode = rr.Response.StatusCode
      errBody = rr.Response.Body
    }
}

if errCode == 0 {
  // yay!
} else if errCode == 404 {
  //  404
} else {
  // 500 with whatever errBody is
}
```

Reply:
> Since we are building and executing the requests I think if we get 400 this should actually be a 500 externally

Hrm, this is not entirely true. At the moment, we pass the traceID as-is and verify only at the querier level. Should we move the traceID validation to the Query Frontend? We'll have to think some about the `/querier/api/traces/{traceID}` endpoint as well, that we want to keep for debugging.

400/401 behaviour will be consistent across all the shards/queriers. If one of them throws 400 all of them will. In that case we will still return 400 after merging responses.

Reply:
This logic makes sense, but one concern is losing the original error code which would make diagnosing a final 500 error harder.  Should there be logging or other here?
```
else if rr.statusCode != 404 {
      errCode = rr.Response.StatusCode
      errBody = rr.Response.Body
    }
```

Reply:
This was fixed as discussed offline. Also PTAL at the new test.